### DISCUSSION TITLE
Oaconvolve

### DISCUSSION THREAD

**Discussion by eleroy** [v923z/micropython-ulab#692] [General]
Hi everyone,

I need to perform fast convolution to do a matched filter on an ESP32.

The signal I have is about 30 000 points with a 100 point filter signal. I can reach about 600 ms per convolution.

Here is my code:
```python
from ulab import numpy as np
import time
import math
def oaconvolve(signal, filter_signal, window_size=2048):
    n=len(filter_signal) # Length of filter
    block_size=window_size-n
    n_blocks = math.ceil(len(signal)/block_size)
    padded_signal = np.zeros(n_blocks*window_size)
    padded_signal[:len(signal)] = signal
    padded_filter = np.zeros(window_size)
    padded_filter[:len(filter_signal)] = filter_signal[::-1]
    fft_filter_signal= np.fft.fft(padded_filter)
    final_signal = np.zeros(len(padded_signal)+n)
    subsignal = np.zeros(window_size)
    for i in range(n_blocks):
        subsignal[:block_size] = padded_signal[i*block_size:(i+1)*block_size]
        a = np.fft.fft(subsignal)
        convolved_block = np.fft.ifft(a*fft_filter_signal)
        final_signal[i*block_size:(i)*block_size+window_size]+=convolved_block
    return final_signal
```

Is there any chance I can make it faster by coding it in C ? Can I use directly the ulab routine in C ?

Thank you

**Comment by v923z** [comment]
One problem with the `fft` in `numpy` is that it doesn't allow you to re-use RAM. This is addressed in https://micropython-ulab.readthedocs.io/en/latest/ulab-utils.html#spectrogram. That alone is not going to help you, because that calculates the absolute value, and not the FFT. However, you can easily change that in 

https://github.com/v923z/micropython-ulab/blob/2b74236c8ce278d876e410c00e90de5bf5e9980e/code/utils/utils.c#L207-L378

So, my suggestion would be to modify the `spectrogram` method of the `utils` module, and use the `scratchpad` and `out` keyword arguments in your implementation of the FFT. That'll save you the overhead of reserving space. 

> Is there any chance I can make it faster by coding it in C ?

`ulab` is written in C, and the FFT routine used is one of the fastest ones. The overhead of `python` is simply the argument inspection at the beginning of the function call, and the memory allocation. But the FFT code can be re-used, because that's already decoupled from `python` here: 

https://github.com/v923z/micropython-ulab/blob/2b74236c8ce278d876e410c00e90de5bf5e9980e/code/numpy/fft/fft_tools.c#L48-L92

**Comment by v923z** [comment]
Since `fft_kernel` is an in-place routine, if you don't need your original data, then you can simply call that, and be done with it. If you end up adding your method, I'd like to ask you to create a PR, because this could be useful in general.

**Comment by eleroy** [comment]
Thank you ! I did succeed to write it in C.

With a signal about 30 000 points and a filter about 200 points

I get this performance:

ulab C OAconvolve 502.704 ms
micropython+ulab OAconvolve 716.302 ms
ulab linear Convolve 3482.501 ms

I'll try to open a PR. I guess I will include the routine oaconvolve in signal.c.

**Reply by v923z** [reply]
Yes, `oaconvolve` is part of `scipy.signal`, so that would be the place for that.

**Reply by v923z** [reply]
> I get this performance:
> 
> ulab C OAconvolve 502.704 ms micropython+ulab OAconvolve 716.302 ms ulab linear Convolve 3482.501 ms

Do you know where the time is used? Because it might actually make more sense to add the `out` keyword to the FFT, and do the rest in `python`, as in your original implementation.

**Reply by eleroy** [reply]
Hi,

I think for small kernel size and large signal size, the fft segments become small and the python iteration becomes slow.

For larger kernel size, the iteration is not prevalent and I guess it's the memory allocation that sets the gap.

**Reply by v923z** [reply]
What I was driving at is the question, where you burn time: you already have an implementation in `python`, and one of the issues with that is that the FFT can't re-use the RAM. The rest won't be significantly faster in C, and that's why I was trying to suggest to fix the FFT code, and leave the rest in `python`. But I would commit to that only, if I knew for sure that not too much time is lost with the `python` loop. Hence my question.

**Comment by eleroy** [comment]
![25](https://github.com/user-attachments/assets/2daebd7f-edf0-4207-b0f8-27a2d64d96f0)
![100](https://github.com/user-attachments/assets/3709ef11-bb46-468f-a929-d861f83165e1)
![200](https://github.com/user-attachments/assets/ba18bd7b-d3e6-456d-8b86-47c29684773a)

Time of computation as a function of kernel size. As kernel gets small the advantage of oaconvolve reduces.

There has been discussion about a C++ OAConvolve implementation https://github.com/scipy/scipy/issues/16802 that might be helpful
